config/src/config/override_node_config.rs

```diff
@@ -0,0 +1,299 @@
+                base_config
+            ),
+        },
+        serde_yaml::Value::Number(override_value) => match base_config {
+            serde_yaml::Value::Number(base_value) => {
+                if override_value == base_value {
+                    Ok(None)
+                } else {
+                    Ok(Some(override_config))
+                }
+            },
+            _ => bail!(
+                "base does not match override: Number({}), {:?}",
+                override_value,
+                base_config
+            ),
+        },
+        serde_yaml::Value::String(override_value) => match base_config {
+            serde_yaml::Value::String(base_value) => {
+                if override_value == base_value {
+                    Ok(None)
+                } else {
+                    Ok(Some(override_config))
+                }
+            },
+            _ => bail!(
+                "base does not match override: String({}), {:?}",
+                override_value,
+                base_config
+            ),
+        },
+        serde_yaml::Value::Sequence(override_value) => match base_config {
+            serde_yaml::Value::Sequence(base_value) => {
+                if override_value == base_value {
+                    Ok(None)
+                } else {
+                    Ok(Some(override_config))
+                }
+            },
+            _ => bail!(
+                "base does not match override: {:?}, {:?}",
+                override_config,
+                base_config
+            ),
+        },
+    }
+}
+
+/// A utility struct for managing a node config that is overriding a base config, and outputting a
+/// yaml representation of it. In most cases, the base config will be the default config, and the
+/// output will be a minimal yaml diff that an aptos node can read.
+///
+/// In rare cases you may want to explicitly write a yaml value that is the same as the default
+/// config (e.g., to avoid the config being optimized by ConfigOptimizer). To do this, change the
+/// base config to a different value before calling get_yaml().
+#[derive(Deserialize, Serialize, PartialEq, Debug, Clone)]
+pub struct OverrideNodeConfig {
+    override_config: NodeConfig,
+    base_config: NodeConfig,
+}
+
+impl OverrideNodeConfig {
+    pub fn new(override_config: NodeConfig, base_config: NodeConfig) -> Self {
+        Self {
+            override_config,
+            base_config,
+        }
+    }
+
+    pub fn new_with_default_base(override_config: NodeConfig) -> Self {
+        Self {
+            override_config,
+            base_config: NodeConfig::default(),
+        }
+    }
+
+    pub fn override_config(&self) -> &NodeConfig {
+        &self.override_config
+    }
+
+    pub fn override_config_mut(&mut self) -> &mut NodeConfig {
+        &mut self.override_config
+    }
+
+    pub fn base_config(&self) -> &NodeConfig {
+        &self.base_config
+    }
+
+    pub fn base_config_mut(&mut self) -> &mut NodeConfig {
```

Comment:
Seems this is used for testing purpose only, if so, mark this test-only ?

Reply:
It's currently only used for testing, but it's not designed only for testing -- if someone needs to write the default value into a yaml they could use this.